Helper methods

ThomasdenH opened the pull request:
This adds some helpful conversion functionality.

`Amount -> CompleteAmount`
`NaiveDate -> Date`
`str -> Flag`

mbudde commented on beancount-core/src/date.rs:
```diff
@@ -0,0 +1,31 @@
+
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct Date<'a> {
+    s: Cow<'a, str>
```
I think this either should be public or there should be some other way of extracting the string from the `Date` struct.

I would also suggest using a tuple struct here instead, the field name is not very descriptive anyway?

ThomasdenH replied:
That's a good point. Isn't the parameter from a tuple struct public anyway?

mbudde commented on beancount-core/src/flags.rs:
```diff
@@ -1,12 +1,27 @@
     }
 }
+
+impl<'a, I: Into<Cow<'a, str>>> From<I> for Flag<'a> {
```
You are using `Into` here unlike in `From<Cow<_>> for Date`. This think those should be consistent. Which is better?

twilco replied:
I think `From` is preferred when possible (it's not always possible), since implementing `From` gets you `Into` for free.

https://doc.rust-lang.org/stable/rust-by-example/conversion/from_into.html

mbudde replied:
What I'm trying to point out is that the `From` impl for `Flag` is generic over `I: Into<Cow<'a, str>>` while the `From` impl for `Date` is not:
```rust
 impl<'a, I: Into<Cow<'a, str>>> From<I> for Flag<'a> {}
 impl<'a> From<Cow<'a, str>> for Date<'a> {}
```
I think these should be consistent since they are both wrappers around strings.

ThomasdenH replied:
That's true. I think that it's better to drop the generics, and potentially add more implementations for `&str` and `String`. The Rust compiler had trouble deciding which implementation to use, but I'm not sure why, as `NaiveDate` doesn't implement `Into<Cow>`.

It doesn't matter much whether to use `Into` or `From`, it's just that when implementing `From` is preferable. In any case, a type has either none,  just `Into` or both.

twilco commented:
> What's the status on travis? It doesn't look like it's running tests for PRs?

Should be all set now.

ThomasdenH commented:
> I'm getting some build errors in beancount-parser because it can't convert &str to Date.

That's odd, not sure why that didn't happen for me.

ThomasdenH commented on beancount-core/src/flags.rs:
```diff
@@ -1,12 +1,27 @@
+            Flag::Warning
+        } else {
+            Flag::Other(s)
+        }
```
The trouble is that it's difficult to match over a `Cow`. I'll try to make it a bit cleaner.

twilco replied:
Ah, right.  Would dereferencing the cow work?

```rust
match &*s {
   "*" | "txn => Flag::Okay,
   ...
}
```

We don't need to belabor this long.  If the above won't work, don't invest too much time into it.

ThomasdenH commented:
I have addressed the suggestions. One thing I don't like too much is having multiple impls of `From` for `Cow`, `&str` and `String`. I thnk explicit conversion is better here though.

twilco commented on beancount-core/src/date.rs:
```diff
@@ -0,0 +1,47 @@
+}
+
+#[cfg(feature = "chrono")]
+impl From<NaiveDate> for Date<'static> {
```
Can this be the anonymous lifetime instead (e.g. `Date<'_>`) rather than the static lifetime?  Or is there a reason the static lifetime was used here?  I believe this change would allow users to use convert dates of any lifetime, rather than only those with a static lifetime.

Same question for the other two static lifetimes in this PR.
